### ISSUE TITLE
Font Overview view(s)

### ISSUE THREAD

**Issue by justvanrossum** [googlefonts/fontra#102]
Font overview:
- grid view
- list view / spreadsheet

Also affects the project list: it needs a way to either jump to a font overview or to a glyph editor.

**Comment by justvanrossum** [comment]
@ollimeier, we have this old issue.

I suggest to for now forget about the list view, and focus on the cell view. A list view will surely be handy, but better to save it for later.

My thinking-aloud notes from our call:

- glyph creation/deletion
- ds axis sliders
- char set selection
- glyph filtering
- smart sets???
- menu bar

**Comment by justvanrossum** [comment]
More notes about implementation:
- Create a new view called "fontoverview" (or "glyphoverview"?)
- Use the `glyph-cell` web component (as used in the "related glyphs" panel), although it may need enhancement. We probably need to make it resizable, it is currently rather static.

Fontra Pak: either it should always go to this new font overview, or we should allow the user to choose.

**Comment by ollimeier** [comment]
Hi @justvanrossum , here are some first sketches:

![Image](https://github.com/user-attachments/assets/5d12646b-85f5-41bd-a26a-f3222b55f502)
![Image](https://github.com/user-attachments/assets/2d5d0b05-056a-49e0-87ba-a3a03bf7e1d4)
![Image](https://github.com/user-attachments/assets/5936f6ce-41f4-490b-a8d7-5138ce12e7fb)
![Image](https://github.com/user-attachments/assets/e5fe6517-db3f-4877-9351-9bd1771781af)
![Image](https://github.com/user-attachments/assets/3f54b9e1-4439-40aa-9380-16038e7f5126)

You can find the images also here:
https://github.com/googlefonts/fontra/tree/issue-102-font-overview/sketches

**Comment by justvanrossum** [comment]
Thanks!

Some quick comments and further thoughts.

I don't think we want "Glyph info" here. I want that column to focus on glyph/character selection and navigation.

The third image shows sections: that is good.

Organizing the cell layout, some possible options:
- sort by XXX
- group by XXX

We need to have a clear concept of how to show:
- what glyphs are in the font
- what glyphs do we need to see (that are possible not yet in the font)
  - based on unicode/script/language?
  - based on customizable lists of characters/glyphs (not sure where such lists should be stored, the could potentially be user-specific or project-specific, to be decided)
  - either way, this should be super dynamic, so the user can switch around desired sets easily

Btw. please don't do sketching in a branch, this has no place in the file history of the project. We don't need to permanently keep track of ephemeral sketches. If we need to share files, there are other ways. For now to paste some sketch images in comments is all we need.

**Comment by justvanrossum** [comment]
I'm struggling with the concepts of "filtering" vs. "sorting" and "grouping".

- I'd like to quickly filter by glyph name
- I'd like to filter by "smallcaps" (`*.sc`), yet maybe keep "group by" settings on top of that

Btw. I think the "sections" should not necessarily be unique sets: a glyph should be able to appear in multiple sections.

Potentially, "sections" could be accordion items, like in the "related glyphs" panel. But I'm not super sure about that. (Although similarities between the font overview and related glyph panels are probably good things.)

**Comment by ollimeier** [comment]
Some general thoughts while spending some time with it:
- I am wondering if the new **glyph-overview**-view is more like the **editor**-view or more like the **fontinfo**-view? The question is, can you imagine to have more panels for the **glyph-overview**-view? For example would a list-view be a separate panel?
- How do we handle the top menu bar without duplicating code? This would require a massive refactoring, because the whole top menu bar is stored in editor.js with a lot of dependencies like `visualizationLayers` to build for example the **Glyph editor appearance** menu dynamically. This sounds like a separate task to me. 

What do you think?

**Comment by justvanrossum** [comment]
- I expect (eventually) a list view to be an in-place alternative for the cell view: You choose the one or the other. (Yet we can have multiple font overview tabs open for a single font)
- Some of the top menu bar's *contents* may have to be duplicated, but I suspect that only a fraction from the editor makes sense in a glyph overview view. It is fine to see the menu bar part of this as a separate task.

**Comment by ollimeier** [comment]
![Image](https://github.com/user-attachments/assets/68efd47d-0b44-4427-b191-56c9f88359f5)
![Image](https://github.com/user-attachments/assets/35c494de-cc0f-4a96-be45-53236eb67868)

**Comment by ollimeier** [comment]
I am not sure about the sidebar idea. When we make it equal to the editor view, we hide glyphs. When you search for a specific glyph, you would not even see it, because it would be behind the sidebar. We could adjust the width of the element with the glyphs, so it's always on the right side of the sidebar.

Within 'Sorting by', I can imaging the following sorting possibilities:
- Unicode ranges
- Glyph ID (imagine you have a font file open)
- Custom sorting
- maybe other presets by designers and engineers (e. all glyphs with the same accents next to each other)

Filtering will be probably the most important one, but also very difficult. 
I imagine, if no checkbox is set to any filter, all glyphs will be displayed. But there might be sub-categories, like:
- Numbers 
   - Basic numbers
   - Fractions
   - Tabular figures
   - Old-style figures
   - etc.

**Comment by justvanrossum** [comment]
> When we make it equal to the editor view

So: it obviously shouldn't be *equal* to the editor view. Also, I don't think we should start with multiple sidebars. Let's have one for now. So no tab item.

The entire view should initially simply be a CSS grid with two columns: left sidebar, glyph cell area.

Font source selector: this should be a popup menu.

**Comment by ollimeier** [comment]
Thanks for the fast feedback! Does this go into the right direction:

![Image](https://github.com/user-attachments/assets/f79afbba-c0a6-412f-b380-cf98a5aca6bd)

The difference here is, that the 'Seach' is below the Font Source selector:
![Image](https://github.com/user-attachments/assets/ab0c55eb-80a3-4862-a61b-2cb1a7d5ec27)

**Comment by justvanrossum** [comment]
> Does this go into the right direction

Yes. I suggest to start implementing the skeleton:

- New "fontoverview" view
- Sidebar, with initially only the "Font source" and "Search glyphs" items
- For now, just one section for all glyphs

This will allow us to discover shortcomings in the glyph-cell component, and will help us flesh out how we should interact with the cell view, in terms of code.

Practical things we need to sort out:

- How to get the unicode category/script/etc properties available in the front end
  - If there is an npm package we should review it
  - We have all the needed data avalable in Python, from `fontTools.unicodedata`, but it is a lot of data
    - We can't do a server call per glyph (that's way too expensive)
    - Downloading _all_ needed data up front may be too much (would cause a delay before we can sort/group/filter)

We have `unicode-utils.js`, that implements the "decomposes as" mapping. We could add things there, if we can keep things small. It is currently under 200kB, but I worry that it will need to become much much bigger to contain all data.

**Comment by ollimeier** [comment]
About 'unicode data as npm package', I found this:
- https://www.npmjs.com/package/unicode-properties

This looks very promising to me, made by devongovett who is responsible for fontkit.js (which I really like). I expect this was also the reason for it, because fontkit.js uses `unicode-properties`.

**Comment by justvanrossum** [comment]
The `unicode-properties` JS lib looks fantastic.

**Comment by justvanrossum** [comment]
> The unicode-properties JS lib looks fantastic.

Except that it hasn't been updated in two years (and unicode-trie hasn't in 5 years), and supports Unicode 12 :( 

It looks seriously unmaintained. In the worst case we should fork it and update to more recent Unicode, but it's hard to tell what's involved.

**Comment by justvanrossum** [comment]
Superseded by #1886
